File: FPTAS.py
import numpy as np
import time

e = 0.01  # epsilon

# Generate a sample.
N = np.random.randint(5, 21)  # number of groups
V = np.random.random() * 800 + 200  # knapsack capacity (volume)
S = [np.random.randint(1, 1001) for _ in range(N)]  # numbers of items of each group
v = [[np.random.random() * 99 + 1 for _ in range(S[i])] for i in range(N)]  # values of each item
w = [[np.random.random() * 99 + 1 for _ in range(S[i])] for i in range(N)]  # weights of each item
x = [[0] * S[i] for i in range(N)]  # solution
# print("N =", N)
# print("V =", V)
# print("S =", S)
# print("v =", v)
# print("w =", w)

# FPTAS
st = time.perf_counter()
P = max([max(v[i]) for i in range(N)])  # max value
K = e * P / N
vs = [[int(v[i][j] / K) for j in range(S[i])] for i in range(N)]  # scaled values
vm = N * int(N / e) + 1  # theoretical max total scaled value + 1
dp = np.inf * np.ones((N + 1, vm))
dp[0, 0] = 0
sol = np.zeros((N, vm), dtype="int")
for i in range(N):
    dp[i + 1, 0: vm] = np.copy(dp[i, 0: vm])
    for k in range(S[i]):
        v_temp = vs[i][k]
        w_temp = w[i][k]
        sol[i, v_temp: vm][dp[i, 0: vm - v_temp] + w_temp < dp[i + 1, v_temp: vm]] = k + 1
        dp[i + 1, v_temp: vm] = np.minimum(dp[i + 1, v_temp: vm], dp[i, 0: vm - v_temp] + w_temp)
oa = 0  # approximate optima
for j in range(vm - 1, -1, -1):
    if dp[N, j] <= V:
        oa = K * j
        # print("opt =", K * j)
        break
op = 0  # primal optima
for i in range(N - 1, -1, -1):
    if sol[i, j] > 0:
        x[i][sol[i, j] - 1] = 1
        op += v[i][sol[i, j] - 1]
        j -= vs[i][sol[i, j] - 1]
# print("X =", x)
et = time.perf_counter()
print("execute time =", et - st)
print("primal optima =", op)
print("approximate optima =", oa)
print("x =", x)

File: FPTAS.cpp
#include <iostream>
#include <vector>
#include <ctime>
using namespace std;

int main()
{
	srand(time(NULL));
	
	double e = 0.01;
	
	// Generate a sample.
//	int N = rand() % 100 + 1;
	int N = 20;  // for testing
	double V = (rand() % 9999 + 1) / 100.0;
	vector<int> S(N);
	vector<vector<double>> v(N), w(N);
	vector<vector<int>> x(N);
	for (int i = 0; i < N; i++) {
//		S[i] = rand() % 100 + 1;
		S[i] = 20;  // for testing
		vector<double> vi(S[i]), wi(S[i]);
		vector<int> xi(S[i], 0);
		v[i] = vi;
		w[i] = wi;
		x[i] = xi;
		for (int j = 0; j < S[i]; j++) {
			v[i][j] = (rand() % 9999 + 1) / 100.0;
			w[i][j] = (rand() % 9999 + 1) / 100.0;
		}
	}
//	cout << "N = " << N << endl;
//	cout << "V = " << V << endl;
//	cout << "S = ";
//	for (int i = 0; i < N; i++)
//		cout << S[i] << ' ';
//	cout << endl;
//	for (int i = 0; i < N; i++) {
//		cout << (i == 0 ? "v = " : "    ");
//		for (int j = 0; j < S[i]; j++)
//			cout << v[i][j] << ' ';
//		cout << endl;
//	}
//	for (int i = 0; i < N; i++) {
//		cout << (i == 0 ? "w = " : "    ");
//		for (int j = 0; j < S[i]; j++)
//			cout << w[i][j] << ' ';
//		cout << endl;
//	}
	
	// FPTAS
	clock_t st = clock();
	double P = 0;
	for (int i = 0; i < N; i++)
		for (int j = 0; j < S[i]; j++)
			if (P < v[i][j])
				P = v[i][j];
	for (int i = 0; i < N; i++)
		for (int j = 0; j < S[i]; j++)
			v[i][j] = int(v[i][j] * N / e / P);
	vector<double> dp(N * int(N / e) + 1, 100);
	dp[0] = 0;
	vector<vector<int>> sol(N);
	for (int i = 0; i < N; i++) {
		vector<int> soli(N * int(N / e) + 1, 0);
		sol[i] = soli;
		for (int j = N * int(N / e); j >= 0; j--)
			for (int k = 0; k < S[i]; k++)
				if (j >= v[i][k] && dp[j - v[i][k]] + w[i][k] < dp[j]) {
					dp[j] = dp[j - v[i][k]] + w[i][k];
					sol[i][j] = k + 1;
				}
	}
	int opt;
	for (opt = N * int(N / e); opt >= 0; opt--)
		if (dp[opt] <= V) {
//			cout << "opt = " << opt * e * P / N << endl;
			break;
		}
	for (int i = N - 1, j = opt; i >= 0; i--)
		if (sol[i][j] > 0) {
			x[i][sol[i][j] - 1] = 1;
			j -= v[i][sol[i][j] - 1];
		}
//	for (int i = 0; i < N; i++) {
//		cout << (i == 0 ? "x = " : "    ");
//		for (int j = 0; j < S[i]; j++)
//			cout << x[i][j] << ' ';
//		cout << endl;
//	}
	clock_t et = clock();
	cout << 1.0 * (et - st) / CLOCKS_PER_SEC << endl;
	return 0;
}
